File: api/src/infra/redis.ts
import { createClient, RedisClientType, RedisFunctions, RedisModules, RedisScripts } from "@redis/client";
import { type Logger } from "winston";

// this function will initialize the redis client and returns the redis instance
const initializeRedis = async (logr: Logger, redisUri: string): Promise<RedisClientType<RedisModules, RedisFunctions, RedisScripts>> => {
  try {
    const client = createClient({ url: redisUri });
    await client.connect();

    logr.info("Connected successfully to redis server");

    // set the dummy key in redis
    client.set("pixelriver", `service-initialized-${Date.now()}`, { EX: 600 });

    return client;
  } catch (error) {
    logr.error(error);
    throw error;
  }
};

export const getRedis = async (logr: Logger): Promise<RedisClientType<RedisModules, RedisFunctions, RedisScripts>> => {
  if (!process.env.REDIS_URI) {
    logr.error("REDIS_URI is not defined");
    throw new Error("REDIS_URI is not defined");
  }

  return await initializeRedis(logr, process.env.REDIS_URI);
};

File: image_processor/image_processor/core/redis.py
import os
import redis
import logging
import time


def initialize_redis(logger: logging.Logger, redis_uri: str) -> redis.Redis:
    """ this function will initialize the redis client and returns the redis instance """
    try:
        client = redis.Redis.from_url(redis_uri)
        client.ping()

        logger.info("Connected successfully to Redis server")

        # Set a dummy key in Redis with a 600-second expiration
        client.set(
            name="pixelriver-image-processor",
            value=f"service-initialized-{int(time.time())}",
            ex=600
        )

        return client
    except Exception as e:
        logger.error(f"Error connecting to Redis: {e}")
        raise


def get_redis(logger: logging.Logger) -> redis.Redis:
    redis_uri = os.getenv("REDIS_URI")
    if not redis_uri:
        logger.error("REDIS_URI is not defined")
        raise ValueError("REDIS_URI is not defined")

    return initialize_redis(logger, redis_uri)
